### PR TITLE
IDEMPIERE-5836: Adding unit test cases for Non stocked expense curren…

### PR DESCRIPTION
**Pull request by dpansheriya**
…t bhavior

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [ ] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [x] I have tested the direct scenario that my code is solving
- [ ] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [x] New and existing unit tests pass locally with my changes
- [x] I have added unit tests that prove my fix is effective or that my feature works
### Documentation
- [ ] I have made corresponding changes to the documentation as follows:
- - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
- - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

**Comment by dpansheriya** [comment]
I added unit testing for existing behavior of Non stocked expense type product.

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java [hunk 1]
```diff
@@ -0,0 +1,771 @@
...
+														.list();
+		for (MProductCategoryAcct categoryAcct : categoryAccts)
+		{
+			categoryAcct.setCostingMethod(MAcctSchema.COSTINGMETHOD_AveragePO);
+			categoryAcct.saveEx();
+		}
+		return category;
+	}
+
+	private void validateCostingForAveragePO(int productID)
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:760]: this should be named as "assertNoAveragePOCost"

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+			
+			//Validate that averagePO costing method should not be created in the cost for the expense type product
+			validateCostingForAveragePO(product.get_ID());
+
+		
+			// Testing cost and stock of the product after completing MR
+			assertEquals(BD_2, cost.getCurrentCostPrice().setScale(2, RoundingMode.HALF_UP), "Unexpected current cost price");
+			
+			assertEquals(BigDecimal.TEN.setScale(2, RoundingMode.HALF_UP), cost.getCurrentQty().setScale(2, RoundingMode.HALF_UP), "Unexpected current Qty");
+			// Test to check that storage should nor created for the expense type product
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:122]: // Assert no storage records created for expense type product

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+			cost.load(getTrxName());
+			validateCostingForAveragePO(product.get_ID());
+			
+			createSOAndShipmentForProduct(product);
+			cost.load(getTrxName());
+
+			// testing cost and stock of the product after completing Shipment
+			assertEquals(BD_2, cost.getCurrentCostPrice().setScale(2, RoundingMode.HALF_UP), "Unexpected current cost price");
+			assertEquals(BigDecimal.TEN.setScale(2, RoundingMode.HALF_UP), cost.getCurrentQty().setScale(2, RoundingMode.HALF_UP), "Unexpected current Qty");
+			// Test to check that storage should nor created for the expense type product
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:200]: // Assert no storage records created for expense type product

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+				assertNull(error, error);
+			}
+
+			// testing cost and stock of the product after Customer return
+			cost.load(getTrxName());
+
+			assertEquals(BD_2, cost.getCurrentCostPrice().setScale(2, RoundingMode.HALF_UP), "Unexpected current cost price");
+			assertEquals(BigDecimal.TEN.setScale(2, RoundingMode.HALF_UP), cost.getCurrentQty().setScale(2, RoundingMode.HALF_UP), "Unexpected current cost Qty");
+			// Test to check that storage should nor created for the expense type product
+			MStorageOnHand[] storages = MStorageOnHand.getAll( Env.getCtx(), product.get_ID(),
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:317]: // Assert no storage records created for expense type product

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+							"Production Status=" + production.getDocStatus());
+
+			fgCost.load(getTrxName());
+
+			cost.load(getTrxName());
+
+			// Test the Cost and stock for the Expense type's Product
+			assertEquals(new BigDecimal("6.00"), cost.getCurrentQty().setScale(2, RoundingMode.HALF_UP), "");
+			assertEquals(BD_2, cost.getCurrentCostPrice().setScale(2, RoundingMode.HALF_UP), "");
+			// Test to check that storage should nor created for the expense type product
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:541]: // Assert no storage records created for expense type product

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+		MProduct product = new MProduct(Env.getCtx(), 0, null);
+		product.setM_Product_Category_ID(DictionaryIDs.M_Product_Category.STANDARD.id);
+		product.setName("testStandardCosting");
+		product.setValue("testStandardCosting");
+		product.setProductType(MProduct.PRODUCTTYPE_ExpenseType);
+		product.setIsStocked(false);
+		product.setIsSold(true);
+		product.setIsPurchased(true);
+		product.setC_UOM_ID(DictionaryIDs.C_UOM.EACH.id);
+		product.setM_Product_Category_ID(category_ID);
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:585]: duplicate setM_Product_Category_ID call

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java [hunk 7]
```diff
@@ -0,0 +1,776 @@
...
+			assertEquals(rLine.getM_InOutLine_ID(), matchPOs[0].getM_InOutLine_ID());
+			assertTrue(matchPOs[0].getQty().compareTo(BigDecimal.TEN) == 0);
+	
+			cost.load(getTrxName());
+
+
+			// testing cost and stock of the product after completing MR
+			assertEquals(BD_2, cost.getCurrentCostPrice().setScale(2, RoundingMode.HALF_UP), "Unexpected current cost price");
+			assertEquals(BigDecimal.TEN.setScale(2, RoundingMode.HALF_UP), cost.getCurrentQty().setScale(2, RoundingMode.HALF_UP), "Unexpected current Qty");
+			// Test to check that storage should nor created for the expense type product
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java:117]: // Assert no storage records created for expense type product

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java [hunk 8]
```diff
@@ -0,0 +1,776 @@
...
+			cost.load(getTrxName());
+
+			// Create SO And Shipment
+			createSOAndShipmentForProduct(product);
+			cost.load(getTrxName());
+
+			// testing cost and stock of the product after completing Shipment
+			assertEquals(BD_2, cost.getCurrentCostPrice().setScale(2, RoundingMode.HALF_UP), "Unexpected current cost price");
+			assertEquals(BD_5, cost.getCurrentQty().setScale(2, RoundingMode.HALF_UP), "Unexpected current Qty");
+			// Test to check that storage should nor created for the expense type product
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java:199]: // Assert no storage records created for expense type product

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java [hunk 9]
```diff
@@ -0,0 +1,776 @@
...
+			assertEquals(DocAction.STATUS_Completed, production.getDocStatus(),
+							"Production Status=" + production.getDocStatus());
+			fgCost.load(getTrxName());
+
+			cost.load(getTrxName());
+
+			// Test the Cost and stock for the Expense type's Product
+			assertEquals(new BigDecimal("6.00"), cost.getCurrentQty().setScale(2, RoundingMode.HALF_UP), "");
+			assertEquals(BD_2, cost.getCurrentCostPrice().setScale(2, RoundingMode.HALF_UP), "");
+			// Test to check that storage should nor created for the expense type product
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java:560]: // Assert no storage records created for expense type product

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java [hunk 10]
```diff
@@ -0,0 +1,776 @@
...
+		MProduct product = new MProduct(Env.getCtx(), 0, null);
+		product.setM_Product_Category_ID(DictionaryIDs.M_Product_Category.STANDARD.id);
+		product.setName("testStandardCosting");
+		product.setValue("testStandardCosting");
+		product.setProductType(MProduct.PRODUCTTYPE_ExpenseType);
+		product.setIsStocked(false);
+		product.setIsSold(true);
+		product.setIsPurchased(true);
+		product.setC_UOM_ID(DictionaryIDs.C_UOM.EACH.id);
+		product.setM_Product_Category_ID(category_ID);
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeStdCostingTest.java:602]: Duplicate setM_Product_Category_ID call

File: org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java (hunk 2 above)
Comment (dpansheriya) [reply org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:122]: @hengsin  line 123 has that assertions.

Comment (hengsin) [reply org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java:122]: I means change the comment text, not the code

Comment (dpansheriya) [reply org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java]: ohk, Corrected and pushed changes.

Comment (dpansheriya) [reply org.idempiere.test/src/org/idempiere/test/costing/NonStockedExpTypeAvgPOCostingTest.java]: Resolved

**Comment by dpansheriya** [comment]
@hengsin all suggested fix was committed. I am also ready with code and that passed this unit test cases. but once this PR unit test cases tested, will commit so that we make sure that PR has not impact

**Review by hengsin — APPROVED** [review]
tested

**Pull request merged**